homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,468 @@
+            destination = self.destination.split(",")
+            origin = self.origin.split(",")
+
+            _LOGGER.debug(
+                "Requesting route for origin: %s, destination: %s, route_mode: %s, mode: %s, traffic_mode: %s",
+                origin,
+                destination,
+                self.route_mode,
+                self.travel_mode,
+                traffic_mode,
+            )
+            response = self._client.car_route(
+                origin, destination, [self.route_mode, self.travel_mode, traffic_mode]
+            )
+            if isinstance(response, herepy.error.HEREError):
+                # Better error message for cryptic no route error codes
+                if any(error in response.message for error in NO_ROUTE_ERRORS):
+                    _LOGGER.error(NO_ROUTE_ERROR_MESSAGE)
+                else:
+                    _LOGGER.error("API returned error %s", response.message)
+                return
+
+            # pylint: disable=no-member
+            route = response.response["route"]
+            summary = route[0]["summary"]
+            waypoint = route[0]["waypoint"]
+            maneuver = route[0]["leg"][0]["maneuver"]
+
+            self.attribution = None
+            self.base_time = summary["baseTime"]
+            if self.travel_mode in TRAVEL_MODES_VEHICLE:
+                self.traffic_time = summary["trafficTime"]
+            else:
+                self.traffic_time = self.base_time
+            distance = summary["distance"]
+            if self.units == CONF_UNIT_SYSTEM_IMPERIAL:
+                # Convert to miles.
+                self.distance = distance / 1609.344
+            else:
+                # Convert to kilometers
+                self.distance = distance / 1000
+            if self.travel_mode in TRAVEL_MODES_VEHICLE:
+                # Get Route for Car and Truck
+                self.route = self._get_route_from_vehicle_maneuver(maneuver)
+            elif self.travel_mode in TRAVEL_MODES_PUBLIC:
+                # Get Route for Public Transport
+                public_transport_line = route[0]["publicTransportLine"]
+                self.route = self._get_route_from_public_transport_line(
+                    public_transport_line
+                )
+            elif self.travel_mode in TRAVEL_MODES_NON_VEHICLE:
+                # Get Route for Pedestrian and Biyclce
+                self.route = self._get_route_from_non_vehicle_maneuver(maneuver)
+            self.origin_name = waypoint[0]["mappedRoadName"]
+            self.destination_name = waypoint[1]["mappedRoadName"]
+
+    @staticmethod
+    def _get_route_from_non_vehicle_maneuver(maneuver: str) -> str:
+        """Extract a Waze-like route from the maneuver instructions."""
+        road_names: List[str] = []
+
+        for step in maneuver:
+            instruction = step["instruction"]
+            try:
+                road_name = instruction.split('<span class="next-street">')[1].split(
```

Comment:
This looks really protocol specific. It should be part of the library.

Reply:
This is an opinionated way on how to construct the route. I can open a PR against the lib but I am not sure it will be accepted.